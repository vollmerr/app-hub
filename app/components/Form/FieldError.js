import styled from 'styled-components';

import theme from 'utils/theme';


const FieldError = styled.p`
  color: ${theme.redDark};
  font-size: 12px;
  margin: 0;
  padding-top: 5px;
`;

export default FieldError;
