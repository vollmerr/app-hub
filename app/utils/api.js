import { delay } from 'redux-saga';
import { call, put } from 'redux-saga/effects';
import decode from 'jwt-decode';
import 'whatwg-fetch';

import { authUserDone } from '../containers/AppHub/actions';


export const DEFAULT_EXPIRE = 1000;
export const TOKEN = 'id_token';


/**
 * Determines if should fetch based off a cache time
 *
 * @param {Date} date             - date of cache
 * @param {Number} cacheSeconds   - number of seconds to cache
 *
 * @return {Bool}                 - if should fetch
 */
export function shouldFetch(date, cacheSeconds = 350) {
  return !date || (new Date(date).getTime() - (cacheSeconds * 1000)) > new Date().getTime();
}


/**
 * Sets a token for authenication in localStorage
 *
 * @param {String} token  - token to set
 */
export function setToken(token) {
  localStorage.setItem(TOKEN, token);
}


/**
 * Gets the authenication token from localStorage
 * @return {String}     - authenication token
 */
export function getToken() {
  return localStorage.getItem(TOKEN);
}


/**
 * Clears the authenication token from localStorage
 */
export function clearToken() {
  localStorage.removeItem(TOKEN);
}


/**
 * Helper function to valdate that token exists and is not expired
 *
 * @param {object} token  - jwt token
 *
 * @return {bool}         - valid ? true : false
 */
export function validToken(token) {
  try {
    const { exp } = decode(token);
    const now = new Date().getTime().toString().substring(0, 10);

    return Number(exp) > Number(now);
  } catch (e) {
    return false;
  }
}


/**
 * Helper function to put jwt token into global redux store
 *
 * @param {String} token    - jwt token
 *
 * @return {Number}         - exipration time
 */
export function* putToken(token) {
  try {
    const {
      sub: sam,
      exp,
      iat,
      ...rest
    } = decode(token);

    const expire = (exp - iat) * 1000; // convert from ms to s

    yield put(authUserDone({ sam, expire, ...rest }));
    return expire;
  } catch (error) {
    throw new Error(error);
  }
}


/**
 * Authenticates a user
 * Checks local storage for token first
 *
 * @param {Object} action   - redux action to dispatch (must have 'type')
 * @param {Number} maxTries - max tries to re attempt authenicating on fail
 */
export function* authenticate(action, maxTries = 3) {
  let expire = DEFAULT_EXPIRE; // try again in 1s by default
  let tries = 0;
  let error = new Error('Failed to authenticate');
  while (tries < maxTries) { // eslint-disable-line
    const token = getToken();
    // if valid local token use that, otherwise get new token
    if (validToken(token)) {
      // update user in global appHub state
      expire = yield call(putToken, token);
      // reset max tries
      tries = 0;
    } else {
      try {
        const options = {
          method: 'get',
          credentials: 'include',
        };
        const { id_token } = yield call(request, API.JWT, options); // eslint-disable-line
        // set into local storage for future authentication caching
        setToken(id_token);
        // update user in global appHub state
        expire = yield call(putToken, id_token);
        // reset max tries
        tries = 0;
      } catch (err) {
        // increment max tries
        tries += 1;
        error = err;
      }
      // max number of tires, set error
      if (tries === maxTries) {
        yield put(authUserDone(error));
      }
    }
    // authenicate again when token has expired
    yield call(delay, expire);
  }
}


/**
 * Parses the JSON returned by a network request
 *
 * @param  {object} response A response from a network request
 *
 * @return {object}          The parsed JSON from the request
 */
export function parseJSON(response) {
  if (response.status === 204 || response.status === 205) {
    return null;
  }
  return response.json();
}


/**
 * Checks if a network request came back fine, and throws an error if not
 *
 * @param  {object} response   A response from a network request
 *
 * @return {object|undefined} Returns either the response, or throws an error
 */
export function checkStatus(response) {
  if (response.status >= 200 && response.status < 300) {
    return response;
  }

  const error = new Error(response.statusText);
  error.response = response;
  throw error;
}


/**
 * Requests a URL, returning a promise
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export function request(url, options) {
  return fetch(url, options)
    .then(checkStatus)
    .then(parseJSON);
}


/**
 * Requests a URL with jwt token in header for authorization, returning a promise
 * Waits until jwt token is set in localStorage before making call
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export function requestWithToken(url, options = {}) {
  let timer = null;

  const requestIfToken = () => { // eslint-disable-line
    // get users jwt token form storage else they dont have one
    const token = localStorage.getItem('id_token') || null;
    // valid token, call api
    if (validToken(token)) {
      clearTimeout(timer);
      // config settings to send api
      const newOptions = {
        method: options.method || 'get',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
        ...options,
      };
      // append optional body to request (POSTs)
      if (options.body) {
        newOptions.body = JSON.stringify(options.body);
      }

      return request(url, newOptions);
    }
    // wait again until token set
    timer = setTimeout(requestIfToken, 100);
  };

  return requestIfToken();
}
