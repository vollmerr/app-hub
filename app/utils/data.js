/**
 * Formats an item object to the correct format
 * for rendering
 *
 * @param {array} item      - item to format
 * @param {string} name     - internal field name for lookup
 * @param {object} field    - field format info
 * @param {object} enums    - display names for enums
 *
 * @return {string}         - formatted item
 */
export function formatItem(item, name, field, enums = {}) {
  let content = item[name];
  // convert all data to array for mapping over
  if (!Array.isArray(content)) {
    content = [content];
  }
  // go through each item, mapping to the correct format
  content = content.map((x) => {
    // is enum mapping
    if (enums[name]) {
      return enums[name][x];
    }
    // is date
    if (field.data && field.data.type === 'DATE') {
      return isNaN(Date.parse(x)) ? '' : new Date(x).toISOString().substr(0, 10);
    }
    return x;
  });
  // join them back as string seperated by commas
  return content.join(', ');
}

/**
 * Formats an array of item objects to the correct format
 * for rendering
 *
 * @param {array} items     - items to format
 * @param {object} fields   - lookup table of fields
 * @param {object} enums    - display names for enums
 *
 * @return {array}          - formatted items as objects of strings
 */
export function formatItems(items, fields, enums) {
  const formattedItems = [];

  items.forEach((item, i) => {
    formattedItems[i] = {};
    Object.keys(item).forEach((k) => {
      formattedItems[i][k] = formatItem(item, k, fields[k], enums);
    });
  });

  return formattedItems;
}

/**
 * Maps a object to the format Office-UI-Fabric-React DetailsList`s expect
 *
 * @param {object} obj    - object to map
 * @param {array} include - keys to include, if empty includes all
 * @param {array} exclude - keys to exclude
 *
 * @return {object}       - mapped object to use in List
 */
export function mapToColumns(obj, include = [], exclude = []) {
  const keys = include.length ?
    include :
    Object.keys(obj);

  return keys
    .filter((key) => !exclude.includes(key))
    .map((key) => ({
      key: obj[key].name,
      name: obj[key].label,
      fieldName: obj[key].name,
      isResizable: true,
      ariaLabel: obj[key].ariaLabel,
      minWidth: obj[key].minWidth || 100,
      maxWidth: obj[key].maxWidth || 200,
      data: obj[key].data || {},
    }));
}
