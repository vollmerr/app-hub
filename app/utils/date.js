export const dateWithoutTime = (date) => {
  if (date && !isNaN(new Date(date))) {
    const d = new Date(date);
    d.setHours(0, 0, 0, 0);
    return d;
  }
  return undefined;
};
