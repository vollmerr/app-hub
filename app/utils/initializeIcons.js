import { registerIcons } from '@uifabric/styling/lib/index';
import appHubIcons from 'images/office365icons.woff';

// http://o365icons.cloudapp.net/
const initializeIcons = () => {
  registerIcons({
    style: {
      MozOsxFontSmoothing: 'grayscale',
      WebkitFontSmoothing: 'antialiased',
      fontStyle: 'normal',
      fontWeight: 'normal',
      speak: 'none',
    },
    fontFace: {
      fontFamily: `"appHubIcons"`, // eslint-disable-line
      src: `url(${appHubIcons}) format('woff')`,
    },
    icons: {
      // NAV
      helpMenu: '\uE006',
      alertsMenu: '\uE010',
      appMenu: '\uE020',
      settingsMenu: '\uE035',
      appHubMenu: '\uE303',
      // __APPS__
      ED: '\uE417',
      Loading: '\uF8FD',
      Demo: '\uF8FD',
      // SPA: '\uE22E',
      Apphub: '\uE300',
      COLD: '\uE408',
      BARS: '\uE259',
      // requried by office-ui
      chevrondown: '\uE088',
      chevronup: '\uE087',
      Search: '\uE039',
      Clear: '\uE353',
      CircleRing: '\uF8FE',
      StatusCircleCheckmark: '\uF8FF',
      sortup: '\uF8FB',
      sortdown: '\uF8FC',
      calendar: '\uE38E',
      up: '\uE27A',
      down: '\uE27B',
      checkmark: '\uF8FA',
    },
  });
};

export default initializeIcons;
