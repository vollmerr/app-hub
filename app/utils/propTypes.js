import PropTypes from 'prop-types';

const { string, bool, shape, func, arrayOf } = PropTypes;

export const metaDataProp = shape({
  title: string,
  desc: string,
  keywords: string,
});

export const routeProp = shape({
  key: string,
  href: string,
  path: string,
  name: string,
});
export const routesProp = arrayOf(routeProp);

// REDUX-FORM
export const inputProp = (value) => shape({
  value,
  onChange: func,
  onBlur: func,
  onFocus: func,
});

export const metaProp = shape({
  error: string,
  touched: bool,
});

export const optionProp = shape({
  key: string,
  text: string,
});
export const optionsProp = arrayOf(optionProp);

