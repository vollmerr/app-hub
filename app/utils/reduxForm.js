export const mapOnBlur = (onBlur, input) => (event, newValue, previousValue) => {
  if (onBlur) {
    onBlur(event, newValue, previousValue);
  }
  input.onBlur(event, newValue, previousValue);
};

export const mapOnFocus = (onFocus, input) => (event) => {
  if (onFocus) {
    onFocus(event);
  }
  input.onFocus(event);
};

export const mapOnChange = (onChange, input) => (event, newValue, previousValue) => {
  if (onChange) {
    onChange(event, newValue, previousValue);
  }
  input.onChange(event, newValue, previousValue);
};
