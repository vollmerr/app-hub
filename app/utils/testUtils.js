import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

/**
 * Tests a styled-component for basic tests
 * such as rending correctly and adopting attributes
 * @param {Object} Component    - React component under test
 * @param {Object} extend       - React component the component under test extends /  inherits from
 */
export function testStyledComponent(Component, extend = null) {
  describe(`<${Component.displayName} />`, () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<Component />);
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it(`should render an <${Component.target}> tag`, () => {
      expect(wrapper.type()).toEqual(Component.target);
    });

    it('should have a className attribute', () => {
      expect(wrapper.prop('className')).toBeDefined();
    });

    it('should adopt a valid attribute', () => {
      const id = 'test';
      wrapper.setProps({ id });
      expect(wrapper.prop('id')).toEqual(id);
    });

    if (extend) {
      it(`should extend ${extend.name}`, () => {
        expect(wrapper.find(extend).exists()).toEqual(true);
      });
    } else {
      it('should not adopt an invalid attribute', () => {
        const attribute = 'test';
        wrapper.setProps({ attribute });
        expect(wrapper.prop('attribute')).toBeUndefined();
      });
    }
  });
}

/**
 * Tests a components for basic mapDispatchToProps information
 * @param {func} mapDispatchToProps   - components mapDispatchToProps function
 * @param {object} actions            - functions in mapDispatchToProps to test
 */
export function testMapDispatchToProps(mapDispatchToProps, actions) {
  const onAction = (action) => (
    `on${action.charAt(0).toUpperCase()}${action.slice(1)}`
  );

  let dispatch;
  let dispatchProps;
  beforeEach(() => {
    dispatch = jest.fn();
    dispatchProps = mapDispatchToProps(dispatch);
  });

  Object.keys(actions).forEach((action) => {
    describe(onAction(action), () => {
      it('should be injected', () => {
        expect(dispatchProps[onAction(action)]).toBeDefined();
      });

      it('should dispatch when called', () => {
        dispatchProps[onAction(action)]();
        expect(dispatch).toHaveBeenCalled();
      });
    });
  });
}
