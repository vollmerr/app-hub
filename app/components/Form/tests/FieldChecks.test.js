import React from 'react';
import { shallow } from 'enzyme';
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';
import { Label } from 'office-ui-fabric-react/lib/Label';
import 'jest-styled-components';

import { testStyledComponent } from 'utils/testUtils';

import Default, { FieldChecks, Check } from '../FieldChecks';
import { FieldChecks as Index } from '../index';
import FieldError from '../FieldError';

testStyledComponent(Check, Checkbox);

describe('Check', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Check errorMessage={null} />);
  });

  it('should render properly when no errorMessage', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render properly when errorMessage', () => {
    wrapper.setProps({ errorMessage: 'test error' });
    expect(wrapper).toMatchSnapshot();
  });
});

const props = {
  label: 'test label',
  name: 'test name',
  meta: {
    error: null,
    touched: false,
  },
  input: {
    value: '',
    onBlur: jest.fn(),
    onFocus: jest.fn(),
    onChange: jest.fn(),
  },
  options: [
    { key: 'key1', text: 'text 1' },
    { key: 'key2', text: 'text 2' },
    { key: 'key3', text: 'text 3' },
  ],
};

describe('FieldChecks', () => {
  let wrapper;
  beforeEach(() => {
    jest.resetAllMocks();
    wrapper = shallow(<FieldChecks {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a `label`', () => {
    expect(wrapper.find(Label).length).toEqual(1);
  });

  it('should render 3 `Check` boxes', () => {
    expect(wrapper.find(Check).length).toEqual(3);
  });

  it('should handle errors if touched', () => {
    const error = 'test error';
    const meta = { error, touched: true };
    wrapper.setProps({ meta });
    expect(wrapper.find(FieldError).length).toEqual(1);
    expect(wrapper.find(FieldError).dive().text()).toContain(error);
  });

  it('should not render errors if not touched or no error', () => {
    let meta = { error: 'test error' };
    wrapper.setProps({ meta });
    expect(wrapper.find(FieldError).length).toEqual(0);

    meta = { touched: true, error: null };
    wrapper.setProps({ meta });
    expect(wrapper.find(FieldError).length).toEqual(0);
  });

  it('should not pass onBlur or onFocus', () => {
    expect(wrapper.find(Check).at(0).prop('onBlur')).toEqual(undefined);
    expect(wrapper.find(Check).at(0).prop('onFocus')).toEqual(undefined);
  });

  it('should be exported (wrapped) in the index', () => {
    expect(Default).toBe(Index);
  });

  describe('componentWillReceiveProps', () => {
    let instance;
    beforeEach(() => {
      instance = wrapper.instance();
    });

    it('should handle setting the options to unchecked when no value from redux-form (on form reset)', () => {
      // start state off with value
      let options = [...props.options];
      options[2].checked = true;
      wrapper.setState({ options });
      // set an input value, should not change options
      let input = { ...props.input, value: 'test value' };
      instance.componentWillReceiveProps({ ...props, input });
      expect(wrapper.state('options')).toEqual(options);
      // simulate clearing form (no input.value), should make all `checked: false`
      options = options.map((option) => ({ ...option, checked: false }));
      input = { ...props.input, value: undefined };
      instance.componentWillReceiveProps({ ...props, input });
      expect(wrapper.state('options')).toEqual(options);
    });
  });

  describe('handleChange', () => {
    let instance;
    let key;
    beforeEach(() => {
      key = props.options[2].key;
      instance = wrapper.instance();
    });

    it('should handle calling redux-form`s onChange', () => {
      // check
      instance.handleChange(null, true, key);
      expect(props.input.onChange).toHaveBeenCalledWith([key]);
      // uncheck
      instance.handleChange(null, false, key);
      expect(props.input.onChange).toHaveBeenCalledWith([]);
    });

    it('should handle updating `options` in the state', () => {
      const options = [...props.options];
      // check
      options[2].checked = true;
      instance.handleChange(null, true, key);
      expect(wrapper.state('options')).toEqual(options);
      // uncheck
      options[2].checked = false;
      instance.handleChange(null, false, key);
      expect(wrapper.state('options')).toEqual(options);
    });

    it('should be called when a checkbox is changed', () => {
      instance.handleChange = jest.fn();
      wrapper.find(Check).at(2).simulate('change', null, true);
      expect(instance.handleChange).toHaveBeenCalledWith(null, true, key);
    });
  });
});
