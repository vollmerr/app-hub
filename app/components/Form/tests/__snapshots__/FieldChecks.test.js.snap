// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldChecks__Check /> should render correctly 1`] = `
<CustomizedCheckbox
  className=""
/>
`;

exports[`Check should render properly when errorMessage 1`] = `
.c0 .ms-Checkbox-checkbox {
  border-color: #a80000;
}

<CustomizedCheckbox
  className="c0"
  errorMessage="test error"
/>
`;

exports[`Check should render properly when no errorMessage 1`] = `
<CustomizedCheckbox
  className=""
  errorMessage={null}
/>
`;

exports[`FieldChecks should render correctly 1`] = `
<div>
  <CustomizedLabel
    label="test label"
  >
    test label
  </CustomizedLabel>
  <FieldChecks__Check
    errorMessage=""
    key="key1"
    label="text 1"
    name="test name"
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "key1",
          "text": "text 1",
        },
        Object {
          "key": "key2",
          "text": "text 2",
        },
        Object {
          "key": "key3",
          "text": "text 3",
        },
      ]
    }
    text="text 1"
  />
  <FieldChecks__Check
    errorMessage=""
    key="key2"
    label="text 2"
    name="test name"
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "key1",
          "text": "text 1",
        },
        Object {
          "key": "key2",
          "text": "text 2",
        },
        Object {
          "key": "key3",
          "text": "text 3",
        },
      ]
    }
    text="text 2"
  />
  <FieldChecks__Check
    errorMessage=""
    key="key3"
    label="text 3"
    name="test name"
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "key1",
          "text": "text 1",
        },
        Object {
          "key": "key2",
          "text": "text 2",
        },
        Object {
          "key": "key3",
          "text": "text 3",
        },
      ]
    }
    text="text 3"
  />
</div>
`;
